keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1228 @@
+
+Types which implement the Go standard `json.Marshaler` interface (`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an `any`, and then rendering that `any` as KYAML.
+
+Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are rendered as per their type.  `MarshalYAML()` is not a Go standard and is not used by KYAML.
+
+<<[UNRESOLVED yaml self-marshaling]>>
+Should we do something for YAML self-marshalling?  It is non-standard and we do not use it in Kubernetes.
+<<[/UNRESOLVED>>
+
+Interfaces are rendered as `null` or their true value type.
+
+Any other type (as per `reflect.Value.Kind()`) will produce an error.
+
+### Multi-line strings
+
+For better or worse, multi-line strings are a thing that people do in Kubernetes, and conventional YAML makes it pretty easy. KYAML has to choose how we want to handle it.  Consider these examples of conventional YAML.
+
+Example 1:
+```
+example_1: |
+    This
+    is a
+    multi-line string
+```
+
+Example 2:
+```
+example_2: |
+    this:
+      is:
+      - embedded
+      - yaml
+      where:
+        whitespace:
+          matters: true
+```
+
+#### Option 1. Treat it like a Go quoted-and-escaped string
+
+This always comes out as one line.
+
+```
+example_1: “This\nis a\nmulti-line string\n”
+example_2: "this:\n  is:\n  - embedded\n  - yaml\n  where:\n    whitespace:\n      matters: true\n"
+```
+
+Pro: Easy and well-defined
+Pro: Can be pasted into most C-like languages to get the original string.
+Con: Hard to read and write by humans when it gets long.
+
+#### Option 2. Use YAML flow-folding and escaped line-breaks
+
+This gets closer to the original without sacrificing too much of the goals of this KEP.
+
+```
+example_1: ”\
+    This\n\
+    is a\n\
+    multi-line string\
+    “
+example_2: "\
+    this:\n\
+    \  is:\n\
+    \  - embedded\n\
+    \  - yaml\n\
+    \  where:\n\
+    \    whitespace:\n\
+    \      matters: true\n\
+    "
+```
+
+Pro: Easier to read than option 1 and close to the original
+Pro: Closer to the original than option 3
+Con: Requires weird EOL handling (`\n\` meaning a newline, plus a line-continuation which  prevents the fold from creating a single leading space on the next line)
+Con: Subtle handling of leading whitespace (all dropped unless escaped)
+
+#### Option 3: Use YAML flow-folding and double line-breaks
+
+This also gets close to the original without sacrificing too much of the goals of this KEP.
+
+```
+example_1: ”\
+    This
+
+    is a
+
+    multi-line string\
+    ”
+example_2: "\
+    this:
+
+    \  is:
+
+    \  - embedded
+
+    \  - yaml
+
+    \  where:
+
+    \    whitespace:
+
+    \      matters: true
+
+    "
+```
+
+Pro: Easier to read than option 1 and close to the original
+Pro: Less escapes than option 2
+Con: Requires weird and surprising EOL handling (double-newline being a directive to YAML to add a single newline to the output)
```

Comment:
We should've also mentioned the surprising behavior of adding a single leading space if you drop one of these lines:

IE:
```yaml
this: "
  is

  - embedded
  "
```
== `this: "is\n- embedded"`
While:
```yaml
this: "
  is
  - embedded
  "
```
==  `this: "is\n - embedded` (notice the space)

I think this is a really bizarre footgun.